AndroidPush/Apps/push-tutorial-two/src/main/java/io/ably/tutorial/push_tutorial_two/MainActivity.java

```diff
@@ -1,13 +1,255 @@
+        LocalBroadcastManager.getInstance(this).registerReceiver(pushReceiver, new IntentFilter(AblyPushMessagingService.PUSH_NOTIFICATION_ACTION));
+    }
+
+    public void performAction(View view) {
+        try {
+            Button button = (Button) view;
+            button.setEnabled(false);
+            String step = button.getText().toString();
+            logMessage("Performing Step: " + step);
+            switch (step) {
+                case STEP_1:
+                    initAblyRuntime();
+                    break;
+                case STEP_2:
+                    initAblyPush();
+                    break;
+                case STEP_3:
+                    sendTestPush();
+                    button.setEnabled(true);
+                    break;
+            }
+        } catch (AblyException e) {
+            logMessage("AblyException " + e.getMessage());
+            handler.sendMessage(handler.obtainMessage(FAILURE));
+        }
+    }
+
+    /**
+     * By default we set client ID as the androidID.
+     *
+     * @return
+     */
+    private String getClientId() {
+        String clientId = Settings.Secure.getString(getContentResolver(), Settings.Secure.ANDROID_ID);
+        return clientId;
+    }
+
+    /**
+     * Step 1: Initialize Ably library and perform authentication with private server
+     *
+     * @throws AblyException
+     */
+    private void initAblyRuntime() throws AblyException {
+
+        ClientOptions options = new ClientOptions();
+        options.environment = BuildConfig.ABLY_ENV;
+        options.key = BuildConfig.ABLY_KEY;
+        options.authUrl = PRIVATE_SERVER_AUTH_URL;
+        options.authParams = new Param[]{new Param("clientId", getClientId())};
+
+        ablyRealtime = new AblyRealtime(options);
+        ablyRealtime.setAndroidContext(getApplicationContext());
+        ablyRealtime.connect();
+
+        ablyRealtime.connection.on(new ConnectionStateListener() {
+            @Override
+            public void onConnectionStateChanged(ConnectionStateChange state) {
+                logMessage("Connection state changed to : " + state.current.name());
+                switch (state.current) {
+                    case connected:
+                        //Go to step 2
+                        handler.sendMessage(handler.obtainMessage(SUCCESS, STEP_2));
+                        break;
+                    case disconnected:
+                    case failed:
+                        handler.sendMessage(handler.obtainMessage(FAILURE));
+                        break;
+                }
+            }
+        });
+    }
+
+    /**
+     * Step 2: Register for Push notification through private server.
+     * We also subscribe for the required channels on Server for the relevant device Id.
+     *
+     * @throws AblyException
+     */
+    private void initAblyPush() throws AblyException {
```

Comment:
This step is supposed to use `activate()`, but specify `customRegistrar = true`. Then handle the intent and use that to get the server to perform the registration.

The reason for doing it that way instead of as you've got it here is that your current code will perform a new registration every time - it has no way of knowing whether or not the push activation state is already initialised. Use the push activation state machine to do that - ie call `activate()`, as you would in the device-based registration case, but handle the intent and then make a request to the server.